### PR TITLE
[SIL] Add [serialized_for_package] to control package-wide resilience domain in Package-CMO.

### PR DESCRIPTION
**Pull request by elsh**
This PR adds a new attribute [serialized_for_package] to allow 
more fine-control over package-wide resilience domain if 
Package CMO is enabled.

The purpose of the attribute includes:
- Indicates that certain types such as loadable types are
allowed in serialized functions in resiliently built module
if the optimization is enabled, which are otherwise disallowed.
- Used during SIL deserialization to determine whether such
functions are allowed.
- Used to determine if a callee can be inlined into a caller
that's serialized without package-cmo, e.g. with an explicit
annotation like @inlinable, where the callee was serialized
due to package-cmo.

Resolves rdar://127870822&127321129&127308331

### CODE REVIEW

**Comment by aschwaighofer** [comment]
We should block inlining a `[serialized] [serialized_for_package]` function into a `[serialized]`-only function until the `[serialized]` attribute is removed. 

Otherwise, you end up with the code originating from a `[serialized_for_package]` function that assumes loadability in a `[serialized]` function (without the `serialized_for_package`) attributed and the verifier will complain.


We achieve the analogous thing for serialized and not-serialized functions here:

https://github.com/apple/swift/blob/8c0726da80dbddbbbf1fc6ba4f0ab39df1ee3e9a/lib/SILOptimizer/Utils/PerformanceInlinerUtils.cpp#L849-L851

https://github.com/apple/swift/blob/8c0726da80dbddbbbf1fc6ba4f0ab39df1ee3e9a/include/swift/SIL/SILFunction.h#L875

**Comment by elsh** [comment]
@swift-ci smoke test

**Comment by slavapestov** [comment]
> We should block inlining a [serialized] [serialized_for_package] function into a [serialized]-only function until the [serialized] attribute is removed.

I think the bits should be mutually exclusive no? So both should not be set at once. [serialized] is the stonger condition so if its set, it implies the same invariants as [serialized_for_package] but adds more.

**Comment by elsh** [comment]
@swift-ci smoke test

**Comment by elsh** [comment]
@swift-ci smoke test

**Comment by elsh** [comment]
@swift-ci smoke test

File: include/swift/SIL/SILLinkage.h [hunk 1]
```diff
@@ -162,6 +162,11 @@ enum IsSerialized_t : unsigned char {
   IsSerialized
 };
 
+enum IsSerializedForPackage_t : unsigned char {
```
Comment (eeckstein) [inline include/swift/SIL/SILLinkage.h:165]: I agree with Slava. You should combine both enums, e.g.
```
enum IsSerialized_t : unsigned char {
  IsNotSerialized,
  IsSerialized,
  IsSerializedForPackage
};
```

**Review by aschwaighofer — APPROVED** [review]
I think we can serialization_t tri-state enum as a follow up PR that also addresses inlinability

**Comment by aschwaighofer** [comment]
we can address ...

**Pull request merged**